components/net_traits/image_cache.rs

```diff
@@ -603,43 +533,104 @@ impl ImageCache {
+            listener.respond(load.image_response.clone());
+            return;
+        }
+        warn!("Couldn't find cached entry for listener {:?}", id);
+    }
 
-    ImageCacheThread::new(ipc_command_sender)
-}
+    /// Inform the image cache about a response for a pending request.
+    pub fn notify_pending_response(&self, id: PendingImageId, action: FetchResponseMsg) {
+        match (action, id) {
+            (FetchResponseMsg::ProcessRequestBody, _) |
+            (FetchResponseMsg::ProcessRequestEOF, _) => return,
+            (FetchResponseMsg::ProcessResponse(_response), _) => {}
+            (FetchResponseMsg::ProcessResponseChunk(data), _) => {
+                debug!("Got some data for {:?}", id);
+                let mut pending_loads = self.pending_loads.lock().unwrap();
+                let pending_load = pending_loads.get_by_key_mut(&id).unwrap();
+                pending_load.bytes.extend_from_slice(&data);
+                //jmr0 TODO: possibly move to another task?
+                if let None = pending_load.metadata {
+                    if let Ok(metadata) = load_from_buf(&pending_load.bytes.as_slice()) {
+                        let dimensions = metadata.dimensions();
+                        let img_metadata = ImageMetadata { width: dimensions.width,
+                                                           height: dimensions.height };
+                        for listener in &pending_load.listeners {
+                            listener.respond(ImageResponse::MetadataLoaded(img_metadata.clone()));
+                        }
+                        pending_load.metadata = Some(img_metadata);
+                    }
+                }
+            }
+            (FetchResponseMsg::ProcessResponseEOF(result), key) => {
+                debug!("Received EOF for {:?}", key);
+                match result {
+                    Ok(()) => {
+                        let bytes = {
+                            let mut pending_loads = self.pending_loads.lock().unwrap();
+                            let pending_load = pending_loads.get_by_key_mut(&id).unwrap();
+                            pending_load.result = Some(result);
+                            debug!("Async decoding {} ({:?})", pending_load.url, key);
+                            pending_load.bytes.mark_complete()
+                        };
+
+                        let (tx, rx) = channel();
+                        thread::Builder::new().name(
+                            "Image decoding async operation".to_owned()).spawn(move || {
+                            let msg = decode_bytes_sync(key, &*bytes);
+                            tx.send(msg).unwrap();
+                        }).expect("Thread spawning failed");
 
-fn decode_bytes_sync(key: LoadKey, bytes: &[u8]) -> DecoderMsg {
-    let image = load_from_memory(bytes);
-    DecoderMsg {
-        key: key,
-        image: image
+                        if let Some(msg) = rx.recv().ok() {
```

Comment:
Synchronously waiting on the result here makes spawning a thread redundant. We should call handle_decoder from inside the thread instead, but that suddenly makes this code more complicated to write. Maybe notify_pending_response needs to accept an additional argument which is an Arc for self? The other option is to given the image cache an interface it can use to enqueue an async operation to call handle_decoder on the original thread; this would mean that there is no risk of an image being decoded halfway through a layout operation, which sounds like a good idea.